src/change_detection.rs

```diff
@@ -103,3 +113,306 @@ impl<T: ?Sized> AsMut<T> for Mut<'_, T> {
         self.deref_mut()
     }
 }
+
+pub struct TraitAdded<'a, Trait: ?Sized + TraitQuery> {
+    marker: PhantomData<&'a Trait>,
+}
+
+pub struct ChangeDetectionFetch<'w, Trait: ?Sized + TraitQuery> {
+    registry: &'w TraitImplRegistry<Trait>,
+    table_ticks: Vec<ThinSlicePtr<'w, UnsafeCell<ComponentTicks>>>,
+    sparse_components: Vec<&'w ComponentSparseSet>,
```

Comment:
I don't like allocating here. Pending benchmarks, its probably faster to just iterate over the registered components and look them up in `fetch`, rather than caching it on the heap. The `All` iterators work this way.

Reply:
@JoJoJet sounds good!